internal/topology/names/names.go

```diff
@@ -0,0 +1,127 @@
+package names
+
+import (
+	"bytes"
+	"fmt"
+	"text/template"
+
+	"github.com/pkg/errors"
+	utilrand "k8s.io/apimachinery/pkg/util/rand"
+)
+
+const (
+	maxNameLength          = 63
+	randomLength           = 5
+	maxGeneratedNameLength = maxNameLength - randomLength
+)
+
+type simpleNameGenerator struct {
+	base string
+}
+
+func (s *simpleNameGenerator) GenerateName() (string, error) {
+	base := s.base
+	if len(base) > maxGeneratedNameLength {
+		base = base[:maxGeneratedNameLength]
+	}
+	return fmt.Sprintf("%s%s", base, utilrand.String(randomLength)), nil
+}
+
+// NameGenerator generates names for objects.
+type NameGenerator interface {
+	// GenerateName generates a valid name. The generator is responsible for
+	// knowing the maximum valid name length.
+	GenerateName() (string, error)
+}
+
+// SimpleNameGenerator returns a NameGenerator which is based on
+// k8s.io/apiserver/pkg/storage/names.SimpleNameGenerator.
+func SimpleNameGenerator(base string) NameGenerator {
+	return &simpleNameGenerator{
+		base: base,
+	}
+}
+
+// ControlPlaneNameGenerator returns a generator for creating a control plane name.
+func ControlPlaneNameGenerator(templateString, clusterName string) NameGenerator {
+	return newTemplateGenerator(templateString, clusterName,
+		map[string]interface{}{})
+}
+
+// MachineDeploymentNameGenerator returns a generator for creating a machinedeployment name.
+func MachineDeploymentNameGenerator(templateString, clusterName, topologyName string) NameGenerator {
+	return newTemplateGenerator(templateString, clusterName,
+		map[string]interface{}{
+			"machineDeployment": map[string]interface{}{
+				"topologyName": topologyName,
+			},
+		})
+}
+
+// MachinePoolNameGenerator returns a generator for creating a machinepool name.
+func MachinePoolNameGenerator(templateString, clusterName, topologyName string) NameGenerator {
+	return newTemplateGenerator(templateString, clusterName,
+		map[string]interface{}{
+			"machinePool": map[string]interface{}{
+				"topologyName": topologyName,
+			},
+		})
+}
+
+// templateGenerator parses the template string as text/template and executes it using
+// the passed data to generate a name.
+type templateGenerator struct {
+	template string
+	data     map[string]interface{}
+}
+
+func newTemplateGenerator(template, clusterName string, data map[string]interface{}) NameGenerator {
+	data["cluster"] = map[string]interface{}{
+		"name": clusterName,
+	}
+	data["random"] = utilrand.String(randomLength)
+
+	return &templateGenerator{
+		template: template,
+		data:     data,
+	}
+}
+
+func (g *templateGenerator) GenerateName() (string, error) {
+	tpl, err := template.New("template name generator").Option("missingkey=error").Parse(g.template)
+	if err != nil {
+		return "", errors.Wrapf(err, "parsing template %q", g.template)
+	}
+
+	var buf bytes.Buffer
+	if err := tpl.Execute(&buf, g.data); err != nil {
+		return "", errors.Wrapf(err, "rendering template: %q", tpl.Name())
```

Comment:
nit: will this read: "rendering template: template name generator: \<error\>". If yes, I would drop tpl.Name()